src/basic_crawler.js

```diff
@@ -173,11 +175,22 @@ class BasicCrawler {
                 return this._isTaskReadyFunction();
             },
             isFinishedFunction: async () => {
-                if (isMaxPagesExceeded() || !this.isRunning) return true;
+                if (!this.isRunning) return true;
+                if (isMaxPagesExceeded()) {
+                    log.info('BasicCrawler: Crawler reached the max requests per crawl limit by crawling '
+                        + `${this.handledRequestsCount} requests and will shut down.`);
+                    return true;
+                }
 
-                return isFinishedFunction
-                    ? isFinishedFunction()
-                    : this._defaultIsFinishedFunction();
+                const isFinished = await this.isFinishedFunction();
+                if (isFinished) {
+                    if (isFinishedFunction) {
```

Comment:
Use of `this.isFinishedFunction` and `isFinishedFunction` is confusing here.

Also I would not bind `this.isFinishedFunction` to `this` as it's not needed anywhere ...

Reply:
also two IFs inside are not easy to read what about something like

```javascript
const isFinished = isFinishedFunction
    ? isFinishedFunction()
    : this._defaultIsFinishedFunction();

if (isFinished) {
    const reason = isFinishedFunction 
        ? 'xxx' 
        : 'yyy';
    log.info(reason);
}

return isFinished;
```

Reply:
Yeah, this is exactly how I had it written earlier, but I felt the two ternaries were confusing...

I'll change it back.